### PR TITLE
C make build

### LINKED ISSUE #22: Add a CMake build

### PR DESCRIPTION
**Pull request by UCLGuichard**
Fixes #22 and partially #20 
I've added a Travis yaml for CI but needs to be configured properly.

### CODE REVIEW

File: src/yaml-parser.cpp [hunk 1]
```diff
@@ -1,79 +1,120 @@
...
+  // Run a batch of methods to parse the YAML and set the
+  // class members accordingly
   this->setParserVariablesFromYaml();
 }
 
-// Method for reading file definition
-void YamlParser::readFile () {
-  YAML::Node config = YAML::LoadFile(this->filename);
+/** 
+    Class method for reading the config file
```
[the remark this answers is not in the document]
Comment (jonc125) [reply src/yaml-parser.cpp]: Also, docstrings should normally go in the header file.

File: src/yaml-parser.h [hunk 2]
```diff
@@ -64,6 +74,6 @@ class YamlParser {
   std::string param1;
   std::string param2;
   
-  YAML::Node config_file;
+  YAML::Node config_file;
```
Comment (jonc125) [inline src/yaml-parser.h]: Avoid hanging whitespace ;)

File: src/yaml-parser.h [hunk 3]
```diff
@@ -6,30 +13,33 @@
...
-  void parseAndSetJ(YAML::Node node);
-  void parseAndSetWProjectionOptions(YAML::Node node);
   void parseAndSetSARA(YAML::Node node);
   std::vector<int> getWavelets(std::string values_str);
   void parseAndSetAlgorithmOptions(YAML::Node node);
-  void parseAndSetPADMM(YAML::Node node);
-  void parseAndSetPD(YAML::Node node);
-
+  
   // Variables
```
Comment (jonc125) [inline src/yaml-parser.h]: As a general rule, avoid public member variables.

File: src/yaml-parser.cpp [hunk 4]
```diff
@@ -1,79 +1,120 @@
...
   this->config_file = config;
 }
 
-// Method for setting the class variables definition
-void YamlParser::setParserVariablesFromYaml () {
-  
+/** 
+    Class method for setting the class members from the YAML inputs
+*/
+void YamlParser::setParserVariablesFromYaml ()
```
Comment (jonc125) [inline src/yaml-parser.cpp]: As noted in #21 I'd prefer not to parse out many member variables within this class. Though that could usefully be a separate PR.

File: src/yaml-parser.h [hunk 5]
```diff
@@ -6,30 +13,33 @@
...
+class YamlParser
+{
  public:
   // Constructor
-  YamlParser(std::string filename);
+  YamlParser(std::string &filepath);
   // Methods
   void readFile();
   void setParserVariablesFromYaml();
   void parseAndSetGeneralConfiguration(YAML::Node node);
```
Comment (jonc125) [inline src/yaml-parser.h]: Many things are still being passed by value, rather than reference-to-const.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+if (NOT yaml-cpp_FOUND)
+   message(FATAL_ERROR "yaml-cpp library not found. Please install.")
+endif (NOT yaml-cpp_FOUND)
+
+add_compile_options(-Wall)
+
+add_library(yaml-parser-lib src/yaml-parser.cpp)
+
+add_executable(parser-test test/parser-test.cpp)
+target_link_libraries(parser-test yaml-parser-lib -lyaml-cpp)
```
[the remark this answers is not in the document]
Comment (UCLGuichard) [reply CMakeLists.txt]: yes of course. done

File: src/Makefile [hunk 7]
```diff
@@ -1,22 +1,22 @@
 yaml-parser:
 	g++ -c yaml-parser.cpp
-test:
-	g++ test.o ../src/yaml-parser.o -lyaml-cpp -lboost_program_options -o test
+parser-test:
```
[the remark this answers is not in the document]
Comment (UCLGuichard) [reply src/Makefile]: No. Removed.

File: src/yaml-parser.cpp [hunk 8]
```diff
@@ -1,79 +1,120 @@
...
-// YamlParser constructor definition
-YamlParser::YamlParser (std::string filename) {
-  // Setting the variable
-  this->filename = filename;
+/** 
+    YamlParser constructor definition
+    
+    @param filepath path to config file
+*/
+YamlParser::YamlParser (std::string& filepath)
```
[the remark this answers is not in the document]
Comment (UCLGuichard) [reply src/yaml-parser.cpp]: Yes, I'm not yet in that const-to-ref C++ routine :). Done

File: src/yaml-parser.cpp [hunk 9]
```diff
@@ -1,79 +1,120 @@
...
+{
+  // Setting the class variable
+  this->filepath = filepath;
```
[the remark this answers is not in the document]
Comment (UCLGuichard) [reply src/yaml-parser.cpp]: Nice, one more C++ trick.

File: src/yaml-parser.h (hunk 2 above)
Comment (UCLGuichard) [reply src/yaml-parser.h]: yes, thanks

File: src/yaml-parser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (UCLGuichard) [reply src/yaml-parser.cpp]: ok. done

File: src/yaml-parser.h (hunk 3 above)
Comment (UCLGuichard) [reply src/yaml-parser.h]: ok, it's done through moving them in a macro.

File: src/yaml-parser.cpp [hunk 10]
```diff
@@ -1,4 +1,4 @@
-/** 
+ /**
```
Comment (jonc125) [inline src/yaml-parser.cpp]: It's more usual not to have a space before the comment here I think.

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,31 @@
+language: c++
+
+
+addons:
+  apt:
+    packages:
+      - build-essential
+      - cmake
+      - libyaml-cpp0.5
```
Comment (jonc125) [inline .travis.yml]: I think you want `libyaml-cpp-dev` instead, so you get the header files and CMake config, not just the library binary.

File: src/yaml-parser.h (hunk 5 above)
Comment (jonc125) [reply src/yaml-parser.h]: So `const YAML::Node& generalConfigNode` and similar in all methods instead of `YAML::Node generalConfigNode` etc.

File: src/yaml-parser.cpp [hunk 12]
```diff
@@ -1,78 +1,92 @@
...
```
Comment (jonc125) [inline src/yaml-parser.cpp]: I don't think it's normal to put a space before a comment like this.

File: src/yaml-parser.cpp (hunk 10 above)
Comment (UCLGuichard) [reply src/yaml-parser.cpp]: ok !

File: .travis.yml (hunk 11 above)
Comment (UCLGuichard) [reply .travis.yml]: I guess I've tried that one ? But maybe worth another attempt. I might think of a better way to check libraries on VMs instead of trying through the .travis.yml.

File: src/yaml-parser.cpp (hunk 12 above)
Comment (UCLGuichard) [reply src/yaml-parser.cpp]: ok !

File: .travis.yml (hunk 11 above)
Comment (UCLGuichard) [reply .travis.yml]: Still failing because I try to `find_package` in `CMake` and it doesn't find the `yaml-cpp` library in Travis.

File: src/yaml-parser.h (hunk 5 above)
Comment (UCLGuichard) [reply src/yaml-parser.h]: Sure, ok :)

File: .travis.yml (hunk 11 above)
Comment (jonc125) [reply .travis.yml]: Looking at https://packages.ubuntu.com/trusty/i386/libyaml-cpp-dev/filelist there is a `/usr/lib/pkgconfig/yaml-cpp.pc` so I'm not sure why CMake is complaining.

Ah, from https://github.com/jbeder/yaml-cpp/issues/336 it's still missing some files which are fixed in the xenial version: https://packages.ubuntu.com/xenial/amd64/libyaml-cpp-dev/filelist, but using [xenial on travis](https://github.com/travis-ci/travis-ci/issues/5821) isn't yet supported except through Docker :(

So building yaml-cpp from source in your `install:` step might be best.

Comment (UCLGuichard) [reply .travis.yml]: Hu, ok ! Thanks for the info, I'll see in that direction.

**Comment by UCLGuichard** [comment]
@jonc125 A quick question, it does kind of work now: a shared yaml-cpp library is created on build but it still fails to find it in the `#include "yaml-cpp/yaml.h"`. Do I have to create a symlink of any kind to make it work ? Thanks.

File: .travis.yml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    - os: linux
+      dist: trusty
+      sudo: required
+      language: generic
+
+before_install:
+
+install:
+  - git clone https://github.com/jbeder/yaml-cpp.git
+  - mv yaml-cpp yaml-cpp-lib
```
Comment (jonc125) [inline .travis.yml]: You can include the folder name to check out as in the `git clone` line

File: src/yaml-parser.cpp [hunk 14]
```diff
@@ -1,78 +1,92 @@
+/** 
+    YamlParser definition for the purify project
+    @author Roland Guichard
+    @version 1.0
+*/
 
+#include <typeinfo>
 #include <iostream>
 #include <fstream>
 #include "yaml-cpp/yaml.h"
```
Comment (jonc125) [inline src/yaml-parser.cpp]: You should normally use `<>` for third-party libraries, not `""`

Comment (UCLGuichard) [reply src/yaml-parser.cpp]: yes, of course.

**Comment by UCLGuichard** [comment]
@jonc125 `make VERBOSE=1` gives no better results:

**Comment by UCLGuichard** [comment]
@jonc125 `make VERBOSE=1` gives no better results:
```
0.16s$ make
Scanning dependencies of target yaml-parser-lib
[ 25%] Building CXX object CMakeFiles/yaml-parser-lib.dir/src/yaml-parser.cpp.o
/home/travis/build/UCLGuichard/yaml-parser-cpp/yaml/src/yaml-parser.cpp:10:27: fatal error: yaml-cpp/yaml.h: No such file or directory
 #include <yaml-cpp/yaml.h>
                           ^
compilation terminated.
make[2]: *** [CMakeFiles/yaml-parser-lib.dir/src/yaml-parser.cpp.o] Error 1
make[1]: *** [CMakeFiles/yaml-parser-lib.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by jonc125** [comment]
I meant putting extra debug output in the CMakeLists so you see what include & library paths it's using to compile the targets. Although `make VERBOSE=1` should also work if you put it on the right make command - when building your code, not building yaml-cpp.

**Comment by UCLGuichard** [comment]
ah, yes stupid me :( wrong place.Ok I'll look in that direction.

**Comment by jonc125** [comment]
So, CMake hasn't added any `-I` flags to set include search paths for the library. Ah, because you're missing a `target_include_directories` directive!

**Comment by UCLGuichard** [comment]
ah ok ! I have so much to learn on CMake...

**Comment by UCLGuichard** [comment]
@jonc125 It works ! However I still don't get why I have to explicitly include the yaml-cpp directory in CMake, what's the purpose of CMake then ? The `find_package` gives you back (at least) two variables: the `INCLUDE_DIR` and `LIBRARY`, the first one is needed to be included to find the header and the second one to the linker, as far as I've understood.

**Comment by jonc125** [comment]
That's correct. CMake does the hard work of figuring out what those variables should be on any given system, providing the ability for users to override its defaults or give hints, and generating the appropriate flags for different compilers. The reason it doesn't magically add them to all targets is that you might not want to link against a given library for all the things you're building, or you might want to do so only in certain circumstances. You can however set things up such that include paths & libraries are shared by all targets defined in a given CMakeLists.txt (or group thereof, possible even subset thereof) so you don't need to set them for each target individually. (Or use macros to semi-automate it.)

**Comment by UCLGuichard** [comment]
Merging for the purpose of integrating to purify. If you have more comments, you can reopen it.
